crates/tap-agent/src/tap/context/receipt.rs

```diff
@@ -208,10 +208,132 @@ impl ReceiptRead<TapReceipt> for TapAgentContext<Horizon> {
 
     async fn retrieve_receipts_in_timestamp_range<R: RangeBounds<u64> + Send>(
         &self,
-        _timestamp_range_ns: R,
-        _receipts_limit: Option<u64>,
+        timestamp_range_ns: R,
+        receipts_limit: Option<u64>,
     ) -> Result<Vec<CheckingReceipt>, Self::AdapterError> {
-        unimplemented!()
+        let receipts_limit = receipts_limit.map_or(1000, |limit| limit);
+
+        let signers = signers_trimmed(self.escrow_accounts.clone(), self.sender)
+            .await
+            .map_err(|e| AdapterError::ReceiptRead {
+                error: format!("{:?}.", e),
+            })?;
+
+        // TODO filter by data_service when we have multiple data services
+
+        let records = sqlx::query!(
+            r#"
+                SELECT 
+                    id,
+                    signature,
+                    allocation_id,
+                    payer,
+                    data_service,
+                    service_provider,
+                    timestamp_ns,
+                    nonce,
+                    value
+                FROM tap_horizon_receipts
+                WHERE
+                    allocation_id = $1
+                    AND payer = $2
+                    AND service_provider = $3
+                    AND signer_address IN (SELECT unnest($4::text[]))
+                AND $5::numrange @> timestamp_ns
+                ORDER BY timestamp_ns ASC
+                LIMIT $6
+            "#,
+            self.allocation_id.encode_hex(),
+            self.sender.encode_hex(),
+            self.indexer_address.encode_hex(),
+            &signers,
+            rangebounds_to_pgrange(timestamp_range_ns),
+            (receipts_limit + 1) as i64,
+        )
+        .fetch_all(&self.pgpool)
+        .await?;
+        let mut receipts = records
+            .into_iter()
+            .map(|record| {
+                let signature = record.signature.as_slice().try_into()
+                    .map_err(|e| AdapterError::ReceiptRead {
```

Comment:
I would love to see these errors being handled via their types instead converting them all to `String`. If you don't want to fix it now, we could create a ticket for that as a separate piece of work.

Reply:
I'd rather create a ticket for that. These decisions were carried over from previous implementation. It was supposed to be some error from `tap_core` that couldn't handle anyhow (because of Clone requirements) and they decided to put String.